### ISSUE TITLE
Crash seen in orchestrator in moveReplicasViaGTID

### ISSUE THREAD

**Issue by sjmudd** [openark/orchestrator#640]
Related to https://github.com/github/orchestrator/issues/353

While I'm still running an older version of orchestrator compared to your code (I intend to address that shortly) I did patch `moveReplicasViaGTID` to add the missing `replicas = RemoveNilInstances(replicas)` line of code.

However, this has not resolved the issue and I'm still seeing a crash:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x88a5b5]

goroutine 9058 [running]:
github.com/github/orchestrator/go/inst.moveReplicasViaGTID.func1.1(0xc0000a32c0, 0xc00404e008)
        /builddir/build/BUILD/orchestrator-3.0.1.9/GO/src/github.com/github/orchestrator/go/inst/instance_topology.go:631 +0x25
github.com/github/orchestrator/go/inst.moveReplicasViaGTID.func1(0xc0000a32c0, 0xc00404e008, 0xc0010c2b00, 0xc00038e620, 0xc00020fa40, 0xc00020fa60, 0xc00020fa80)
        /builddir/build/BUILD/orchestrator-3.0.1.9/GO/src/github.com/github/orchestrator/go/inst/instance_topology.go:651 +0xd3
created by github.com/github/orchestrator/go/inst.moveReplicasViaGTID
        /builddir/build/BUILD/orchestrator-3.0.1.9/GO/src/github.com/github/orchestrator/go/inst/instance_topology.go:630 +0x2b7
```

First: The logging that I saw basically showed the crash occur in at the following line in your current code: https://github.com/github/orchestrator/blob/master/go/inst/instance_topology.go#L628, which is the code which does:

```
defer func() { barrier <- &replica.Key }()
```

I looked at the code again and have a couple of questions which perhaps you can clarify on current behaviour ahead of me suggesting further changes.

I tried to double check ahead of entering the `for` loop if replica was nil. It didn't trigger an issue. I'm tempted to think therefore that there may be some issue of variable scoping with go routines or defer behaviour coming into play.

In any case a couple of questions came up:
* Relocation with GTID should just work. So why not run in parallel? You have a mutex to ensure that only a single relocation happens at once and a barrier to ensure each replica is handled one by one.
* to me I'd expect that GTID relocation with `auto_position = 1` should quite happily run in parallel and there should be no need for `replicaMutex` and it may well be safe to use

```
barrier := make(chan struct{}, len(replicas))
```

* barrier is never read from, so it also seems better to use the construct `struct {}{}` rather than the `replica.Key` which was giving me the error: `panic: runtime error: invalid memory address or nil pointer dereference` avoiding any possibility of the current error being possible.

I'd appreciate your thoughts on these few points as I think a cleaner patch can be offered which should make the problem go away but I may be missing something and hence the questions.

Related (and can go into a separate issue if that keeps things separate and is better):

The function `canMoveViaGTID` may need to be smarter as it currently does not take into account some issues which can arise and which may break replication when used from `moveReplicasViaGTID`. They are:

* erroneous transactions on the new master. if `auto_position = 1` then the slave will try to sync any GTIDs it is missing. Often erroneous transactions will have disappeared from binlogs triggering the sync between the slave and it's new master to fail. So recognising if all needed GTIDs are available would be good.
* moving a delayed slave which has a considerable amount of delay (and thus relay logs) may also cause issues:
** if the equivalent binlogs are on the new master they need to be fetched again which is wasteful but may be unavoidable
** if the equivalent binlogs are not on the new master then replication will break as the `CHANGE MASTER` command will remove any relay logs prior to reconnecting to the new master.
** it seems wise to check `gtid_executed` and `gitd_purged` on the new master with `Retrieved_Gtid_Set` and `Executed_Gtid_Set` from the slave to identify if this might happen and *not* attempt the relocation in this case, providing a helpful error message why this isn't possible

I've bumped into both scenarios so thought I'd make this visible in this issue to see if you had thoughts on this.   Maybe these issues have been solved in newer versions of Orchestrator but I'm not sure that is the case.

Regards,

Simon

**Comment by sjmudd** [comment]
I'm basically running this on my systems right now: https://gist.github.com/sjmudd/c9a5eebb67f9c92acb36b97f917aa4e2 and haven't seen any issues so far but need to wait a little further to be sure.

**Comment by shlomi-noach** [comment]
Some answers, and please see #641 as suggested solution.

Let's begin with the simple thing: `replicaMutex` is required, and does not affect the concurrency of replica relocation: these are executed concurrently. But appending to the shared resources `movedReplicas`, `unmovedReplicas`, `errs` -- all of which are golang arrays, must be synchronized and protected. This synchronization happens _after_ having moved replicas, so it does not affect the move=concurrency.

> I tried to double check ahead of entering the for loop if replica was nil. It didn't trigger an issue. I'm tempted to think therefore that there may be some issue of variable scoping with go routines or defer behaviour coming into play.

The fact `&replica.Key` threw an error is disturbing. `defer` evaluated values upon definition. `replica` could not have been `nil` upon definition. It _might_ turn into nil upon 

https://github.com/github/orchestrator/blob/9b313d00d338bf4c6be71185a115dd710c20725e/go/inst/instance_topology.go#L632

But again, that should not affect the `defer` point of time.

But -- you are right:

> barrier is never read from,

And I'm OK to not spend much time over this. Also, instead of using a `struct{}` let's just use a `sync.WaitGroup`... #641 does that.

Re: `canMoveViaGTID`, I'll answer later.

**Comment by sjmudd** [comment]
Thanks for the rapid feedback: I had overlooked the `movedReplicas`, `unmovedReplicas`, `errs` slices and changes being made to them requiring protection. A comment might have made that more obvious, but maybe I should have seen it.

I did double check documentation about `defer()` behaviour but couldn't see an obvious reason why I saw the nil pointer dereference in this specific case. However, the change to use https://github.com/github/orchestrator/pull/641 looks good as it avoids the problem entirely and also uses standard golang primitives which is simpler.
